code/modules/minimap/minimap_ui.dm

```diff
@@ -0,0 +1,201 @@
+/atom/movable/minimap_ui_handler
+	var/static/max_minimap_id = 0
+	var/minimap_id
+	var/list/viewers = list()
+	var/atom/movable/screen/handler
+	var/obj/minimap
+
+	var/tgui_title
+	var/tgui_theme
+
+	New(parent, control_id = null, obj/minimap, tgui_title, tgui_theme)
+		. = ..()
+		START_TRACKING
+		src.loc = parent
+		src.tgui_title = tgui_title
+		src.tgui_theme = tgui_theme
+
+		if (isnull(control_id))
+			control_id = "minimap_ui_[max_minimap_id++]"
+		src.minimap_id = "[control_id]"
+
+		src.handler = new
+		src.handler.plane = 0
+		src.handler.mouse_opacity = 0
+		src.handler.screen_loc = "[src.minimap_id]:1,1"
+
+		src.minimap = minimap
+		src.minimap.screen_loc = "[src.minimap_id]:1,1"
+		src.handler.vis_contents += src.minimap
+
+	disposing()
+		STOP_TRACKING
+		for (var/client/viewer in src.viewers)
+			if (viewer)
+				viewer.screen -= src.handler
+				viewer.screen -= src.minimap
+		. = ..()
+
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "Minimap")
+			ui.open()
+
+	ui_static_data(mob/user)
+		. = list(
+			"title" = src.tgui_title,
+			"theme" = src.tgui_theme,
+			"minimap_id"= src.minimap_id
+		)
+
+	ui_data(mob/user)
+		src.add_client(user?.client)
```

Comment:
side effects in `ui_data` is pretty jank, can't you just put this in `ui_interact`?

Reply:
Necessary, I think. The skeleton of `minimap_ui` was inspired by `datum/character_preview/multiclient`.